src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs

```diff
@@ -0,0 +1,34 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System;
+
+internal static class TimeExtensions
+{
+    public static long ToTicks(this TimeSpan timeSpan, TimeProvider timeProvider)
+        => timeSpan.ToTicks(timeProvider.TimestampFrequency);
+
+    public static long ToTicks(this TimeSpan timeSpan, long tickFrequency)
+    {
+        if (timeSpan == TimeSpan.MaxValue)
+        {
+            return long.MaxValue;
+        }
+        if (timeSpan == TimeSpan.MinValue)
+        {
+            return long.MinValue;
+        }
+        // The tick frequency should be equal or greater than TicksPerSecond
+        return timeSpan.Ticks * (tickFrequency / TimeSpan.TicksPerSecond);
```

Comment:
`long` division can lose accuracy if the tick frequency is fractionally different than TimeSpan.TicksPerSecond. [`TimeProvider.GetElapsedTime` uses `double` or the division](https://github.com/dotnet/runtime/blob/f73356c1f4b9b292dc4fb02f1dd47e4a61e423f5/src/libraries/Common/src/System/TimeProvider.cs#L113).

```suggestion
        return (long)(timeSpan.Ticks * ((double)tickFrequency / TimeSpan.TicksPerSecond));
```
It's weird to me that we special case min and max value, but otherwise let this potentially overflow. We usually call `ToTicks()` on user-provided timespans like configurable timeouts, right? That means it should be called fairly infrequently. If so, I wonder if we should use checked math to be cautious.

```suggestion
        checked
        {
            return (long)(timeSpan.Ticks * ((double)tickFrequency / TimeSpan.TicksPerSecond));
        }
```

I see `TimeProvider` doesn't bother with checked math for this. I'm not sure why, but it seems like it'd be really bad if this overflowed. We could special case `tickFrequency == TimeSpan.TicksPerSecond` to avoid the checked math in almost all real scenarios. @tarekgh

Reply:
tickFrequency only matches TimeSpan.TicksPerSecond on Windows, not Linux.

I special cased MaxValue because it's used in tests and because customers are likely to use it when they think they're disabling timeouts.

I'll harden this code path.